lib/promscrape/discovery/kuma/api.go

```diff
@@ -0,0 +1,167 @@
+package kuma
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/logger"
+	"io"
+	"net/http"
+	"net/url"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape/discoveryutils"
+	"github.com/VictoriaMetrics/metrics"
+)
+
+var configMap = discoveryutils.NewConfigMap()
+
+type apiConfig struct {
+	client *discoveryutils.Client
+	path   string
+
+	cancel        context.CancelFunc
+	targetsMutex  sync.RWMutex
+	targets       []kumaTarget
+	latestVersion string
+	latestNonce   string
+
+	fetchErrors *metrics.Counter
+	parseErrors *metrics.Counter
+}
+
+const (
+	xdsApiVersion      = "v3"
+	xdsRequestType     = "discovery"
+	xdsResourceType    = "monitoringassignments"
+	xdsResourceTypeUrl = "type.googleapis.com/kuma.observability.v1.MonitoringAssignment"
+	discoveryNode      = "victoria-metrics"
+)
+
+func getAPIConfig(sdc *SDConfig, baseDir string) (*apiConfig, error) {
+	v, err := configMap.Get(sdc, func() (interface{}, error) { return newAPIConfig(sdc, baseDir) })
+	if err != nil {
+		return nil, err
+	}
+	return v.(*apiConfig), nil
+}
+
+func newAPIConfig(sdc *SDConfig, baseDir string) (*apiConfig, error) {
+	ac, err := sdc.HTTPClientConfig.NewConfig(baseDir)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse auth config: %w", err)
+	}
+	parsedURL, err := url.Parse(sdc.Server)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse kuma_sd server URL: %w", err)
+	}
+	apiServer := fmt.Sprintf("%s://%s", parsedURL.Scheme, parsedURL.Host)
+
+	proxyAC, err := sdc.ProxyClientConfig.NewConfig(baseDir)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse proxy auth config: %w", err)
+	}
+	client, err := discoveryutils.NewClient(apiServer, ac, sdc.ProxyURL, proxyAC)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create HTTP client for %q: %w", apiServer, err)
+	}
+
+	apiPath := path.Join(
+		parsedURL.RequestURI(),
+		xdsApiVersion,
+		xdsRequestType+":"+xdsResourceType,
+	)
+
+	cfg := &apiConfig{
+		client:      client,
+		path:        apiPath,
+		fetchErrors: metrics.GetOrCreateCounter(fmt.Sprintf(`promscrape_discovery_kuma_errors_total{type="fetch",url=%q}`, sdc.Server)),
+		parseErrors: metrics.GetOrCreateCounter(fmt.Sprintf(`promscrape_discovery_kuma_errors_total{type="parse",url=%q}`, sdc.Server)),
+	}
+
+	// start updating targets with a long polling in background
+	cancel, initCh := initAndRepeatWithInterval(*SDCheckInterval, func(ctx context.Context) {
```

Comment:
I'd recommend to add config init method startWatcher 

 with possible signature:

```go
func (c* cfg) startWatcher(ctx){
    
    // blocking init
    if err := c.updateTargets(ctx); err != nil {
           logger.Error
     }
    // start background update
    t := time.NewTicker()
    c.wg.Add(1)
    go func() { 
          defer c.wg.Done()
          defer t.Stop()
         for { 
            select 
              case <-t.C { if err := c.updateTargets(ctx); err != nil {} }
             case <- ctx.Done() }  }
  
   }
```

 It should be a bit more readable and prevent resource leak.

Reply:
I rewrote it in this style, but wg wasn't useful here

Reply:
wg should be used at `mustStop` function. E.g.

```go
func (c* cfg) startWatcher(ctx){
    // blocking init
    ...
    // start background work
    c.wg.Add(1)
    go func() { 
          defer c.wg.Done()
          ...
    }
}
func (cfg *apiConfig) mustStop() {
	cfg.cancel()
	cfg.client.Stop()
        cfg.wg.Wait()
}
```

Reply:
or stopped chan struct. Personally, I prefer wg.

Reply:
There it is already done by closing the context by calling the cfg.cancel function, that finishes the goroutine.